# Recommender Systems - Data Mining Lab 2

In [2]:
import scipy
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sys

### Data loading

In [3]:
train_file_path = 'data/lab2_train.csv' 
test_file_path = 'data/lab2_test.csv'

train_data = pd.read_csv(train_file_path, delimiter=',')
test_data = pd.read_csv(test_file_path, delimiter=',')


In [68]:
print(train_data)
print(like_train_data)
like_train_data = train_data[train_data.is_like == True]
like_train_data = like_train_data.sort_values("user_to_id")
match_train_data = train_data[train_data.is_match ==True]
match_train_data = match_train_data.sort_values("user_from_id")

liked_user_count = like_train_data["user_to_id"].nunique()
user_count = like_train_data["user_from_id"].nunique()
match_count = match_train_data["user_from_id"].nunique()

print(user_count)
print(liked_user_count)
print(match_count)

print(like_train_data[like_train_data.user_to_id == 36])
print(like_train_data[like_train_data.user_from_id == 36])


#print(like_train_data)

print(match_train_data)

       user_from_id  user_to_id  is_like is_match
0              1136        3141    False    False
1              2424        3174    False    False
2              1300        3590    False    False
3               800        2736    False    False
4               883         437    False    False
...             ...         ...      ...      ...
76387          2376        3057    False    False
76388          1163         933    False    False
76389          2770        3324    False    False
76390           879         785    False    False
76391           291         470    False    False

[76392 rows x 4 columns]
       user_from_id  user_to_id  is_like is_match
50598            93           2     True    False
46348            24           2     True    False
11643            88           2     True    False
61229            82           2     True    False
71280            67           2     True    False
...             ...         ...      ...      ...
32426          1220     

## Familiarization
### Important properties of the data
There are only 1707 user who have one like or more of the 2107 users who gave likes. Of all of these users there are only 415 matches. Of these matches there are 345 unique users who receive a match. So there is only a small group of users who receive more than one match. And a small number of matches on the total of 76392 likes and dislikes.

The Data is therefore very sparse. Of these 76392 likes and dislikes there are only 12637 likes. This is 16,5% likes. Of these likes there are only 415 matches. Thus only 3,3% mutual likes. This is sparse data to work with.   

### Types of people

- Someone who receives a lot of likes but likes almost no one. User 2 receives 45 likes but likes none of them. Therefore not leading to a match

- Someone who receives and gives about only dislikes. For example user 28 is very active on the platform but gives about only dislikes and receives only dislikes. Therefore not leading to a match

- Someone who gives a lot of likes and receives only dislikes. For example user 18 gives 26 likes and receives no likes back and has thus not a match.

- Someone who receives a lot of likes and gives a lot of likes. For example user 36 gives a lot of likes and receives a lot of likes and eventualy leading to a match.